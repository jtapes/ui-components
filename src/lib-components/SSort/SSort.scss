@import '../../theme/styles/variables';
@import '../../theme/styles/mixins';
@import '../../theme/styles/reset';
@import '../../theme/styles/fonts';
@import '../../theme/styles/global';

.s-sort {
  display: flex;

  &__icon {
    transition: transform $animate--time;

    &--desc {
      transform: rotateX(180deg);
    }
  }
}
