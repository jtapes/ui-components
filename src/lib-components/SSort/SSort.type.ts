export type SortItem = {
  field: string
  label: string
}
export type SortValue = {
  type: string
  direct: string
}
