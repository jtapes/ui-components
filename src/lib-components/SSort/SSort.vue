<template>
  <div class="s-sort">
    Сортировать
    <a
      v-for="(item, index) in sortItems"
      :key="index"
      class="link d-f ai-c ml-24"
      href="#"
      @click.prevent="handleValueChange(item)"
    >
      {{ item.label }}
      <svg-icon
        v-show="value.type === item.field"
        name="sort-16"
        class="icon-16 ml-8 s-sort__icon"
        :class="`s-sort__icon--${value.direct}`"
      />
    </a>
  </div>
</template>

<script lang="ts">
import { Component, Vue, Prop, Model } from 'vue-property-decorator'
import SvgIcon from '@/lib-components/SvgIcon/SvgIcon.vue'
import { SortItem, SortValue } from './SSort.type'

@Component({
  components: {
    SvgIcon,
  }
})
export default class SSort extends Vue {
  @Prop(Array)
  sortItems!: SortItem[]

  @Model('change')
  value!: SortValue

  handleValueChange(item: SortItem): void {
    const newValue = {
      type: item.field,
      direct: 'asc',
    }

    if (item.field === this.value.type) {
      newValue.direct = this.value.direct === 'asc' ? 'desc' : 'asc'
    }

    this.$emit('change', newValue)
  }
}
</script>
