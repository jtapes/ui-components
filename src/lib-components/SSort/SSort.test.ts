import { mount } from '@vue/test-utils'
import SSort from '@/lib-components/SSort/SSort.vue'

describe('SSort', () => {
  test('Is Vue snapshot', () => {
    const wrapper = mount(SSort, {
      propsData: {
        value: {
          field: 'price',
          method: 'desc',
        },
        sortItems: [
          {
            label: 'по цене',
            field: 'price',
          },
          {
            label: 'по количеству',
            field: 'amount',
          },
          {
            label: 'по размеру',
            field: 'length',
          },
        ],
      },
    })

    expect(wrapper.element).toMatchSnapshot()
  })
  test('Is emitted by $emit', async () => {
    const wrapper = mount(SSort, {
      propsData: {
        value: {
          field: 'price',
          method: 'desc',
        },
        sortItems: [
          {
            label: 'по цене',
            field: 'price',
          },
          {
            label: 'по количеству',
            field: 'amount',
          },
          {
            label: 'по размеру',
            field: 'length',
          },
        ],
      },
    })

    const sortItem = wrapper.find('.link')
    await sortItem.trigger('click')
    expect(wrapper.emitted('change')).toBeTruthy()
  })
})
