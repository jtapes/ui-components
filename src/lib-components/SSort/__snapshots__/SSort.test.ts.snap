// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SSort Is Vue snapshot 1`] = `
<div
  class="s-sort"
>
  
  Сортировать
  
  <a
    class="link d-f ai-c ml-24"
    href="#"
  >
    
    по цене
    
    <svg
      class="svg-icon icon-16 ml-8 s-sort__icon svg-icon--sort-16 s-sort__icon--undefined"
      style="display: none;"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        href="#sort-16"
        xlink:href="#sort-16"
      />
    </svg>
  </a>
  <a
    class="link d-f ai-c ml-24"
    href="#"
  >
    
    по количеству
    
    <svg
      class="svg-icon icon-16 ml-8 s-sort__icon svg-icon--sort-16 s-sort__icon--undefined"
      style="display: none;"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        href="#sort-16"
        xlink:href="#sort-16"
      />
    </svg>
  </a>
  <a
    class="link d-f ai-c ml-24"
    href="#"
  >
    
    по размеру
    
    <svg
      class="svg-icon icon-16 ml-8 s-sort__icon svg-icon--sort-16 s-sort__icon--undefined"
      style="display: none;"
      xmlns="http://www.w3.org/2000/svg"
    >
      <use
        href="#sort-16"
        xlink:href="#sort-16"
      />
    </svg>
  </a>
</div>
`;
